File: map-gen-script/src/map_generator/topologies/symmetrical_islands.py
# src/map_generator/topologies/symmetrical_islands.py

import copy
import random
from .base_topology import BaseTopology
from src.map_generator.models.path_info import PathInfo, Coord

class SymmetricalIslandsTopology(BaseTopology):
    """
    Tạo ra các khu vực (đảo) có cấu trúc giống hệt nhau, được ngăn cách
    bởi các khoảng trống hoặc hành lang.
    
    Đây là dạng map lý tưởng để dạy về việc nhận biết mẫu lặp lại
    và phân rã vấn đề thành các hàm có thể tái sử dụng.
    """

    def _create_island_pattern(self, top_left_corner: Coord, pattern_name: str) -> tuple[list[Coord], tuple[int, int]]:
        """
        Tạo ra một mẫu hình hòn đảo cụ thể.
        Hàm này có thể được thay đổi để tạo ra các mẫu phức tạp hơn.
        
        Args:
            top_left_corner (Coord): Tọa độ góc trên bên trái của hòn đảo.
            pattern_name (str): Tên của mẫu hình ('u_shape', 'l_shape', 'plus_shape', 'square_shape').
            
        Returns:
            tuple[list[Coord], tuple[int, int]]: Một tuple chứa:
                - Danh sách các tọa độ tạo nên con đường trên đảo.
                - Kích thước (width, depth) của mẫu hình đó.
        """
        x, y, z = top_left_corner
        path = []
        size = (0, 0)
        
        if pattern_name == 'l_shape':
            # Hình chữ L, kích thước 3x3
            path = [(x, y, z), (x, y, z + 1), (x, y, z + 2), (x + 1, y, z + 2), (x + 2, y, z + 2)]
            size = (3, 3)
        elif pattern_name == 'plus_shape':
            # Hình dấu cộng, kích thước 3x3
            path = [(x + 1, y, z), (x, y, z + 1), (x + 1, y, z + 1), (x + 2, y, z + 1), (x + 1, y, z + 2)]
            size = (3, 3)
        elif pattern_name == 'square_shape':
            # Hình vuông 3x3 rỗng ruột
            path = [
                (x, y, z), (x + 1, y, z), (x + 2, y, z),
                (x + 2, y, z + 1), (x + 2, y, z + 2),
                (x + 1, y, z + 2), (x, y, z + 2), (x, y, z + 1)
            ]
            size = (3, 3)
        else: # Mặc định là 'u_shape'
            # Tạo một mẫu hình chữ U đơn giản, kích thước 2x2
            # Path: (x,z) -> (x+1,z) -> (x+1,z+1) -> (x,z+1)
            path = [(x, y, z), (x + 1, y, z), (x + 1, y, z + 1), (x, y, z + 1)]
            size = (2, 2)
        
        return path, size

    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể với số lượng đảo khác nhau.
        """
        count = 0
        # Lặp qua các số lượng đảo có thể có
        for num_islands in range(2, 6):
            if count >= max_variants: return
            
            variant_params = copy.deepcopy(params)
            variant_params['num_islands'] = num_islands
            yield self.generate_path_info(grid_size, variant_params)
            count += 1

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        """
        Tạo ra một chuỗi các hòn đảo giống hệt nhau.

        Args:
            params (dict): Có thể chứa 'num_islands' để tùy chỉnh số lượng đảo.

        Returns:
            PathInfo: Một đối tượng chứa thông tin về tất cả các hòn đảo.
        """
        print("    LOG: Generating 'symmetrical_islands' topology...")
        
        # --- (CẢI TIẾN) Đọc và ngẫu nhiên hóa số lượng đảo ---
        num_islands_param = params.get('num_islands', 2)
        if isinstance(num_islands_param, list) and len(num_islands_param) == 2:
            # Nếu num_islands là một khoảng [min, max], chọn ngẫu nhiên một giá trị
            num_islands = random.randint(num_islands_param[0], num_islands_param[1])
        else:
            # Nếu không, sử dụng giá trị được cung cấp hoặc mặc định
            num_islands = num_islands_param
            
        # [MỚI] Đọc loại mẫu hình đảo từ params
        island_pattern_name = params.get('island_pattern', 'u_shape')

        # --- [REWRITTEN] Logic tính toán khoảng cách giữa các đảo ---
        # Ưu tiên đọc khoảng cách từ params, nếu không có thì tự động tính toán.
        _, island_size = self._create_island_pattern((0,0,0), island_pattern_name)
        island_width = island_size[0]

        if 'island_gap' in params:
            # Lấy số ô trống từ params
            island_gap = int(params['island_gap'])
            # Khoảng cách giữa các điểm bắt đầu = chiều rộng đảo + số ô trống
            island_spacing = island_width + island_gap
        else:
            # Tự động tính toán khoảng cách tối ưu để giãn đều
            total_islands_width = num_islands * island_width
            available_spacing = grid_size[0] - total_islands_width - 2 # Trừ 2 cho lề an toàn
            
            # Nếu chỉ có 1 đảo, spacing không có ý nghĩa.
            # Khoảng cách giữa các góc bắt đầu của hai đảo liên tiếp
            island_spacing = island_width + (available_spacing // (num_islands - 1)) if num_islands > 1 else 0 # type: ignore

        # Tính toán vị trí bắt đầu an toàn
        start_x = 1 
        start_z = random.randint(1, grid_size[2] - island_size[1] - 2) # Trừ kích thước đảo và lề
        y = 0
        
        start_pos: Coord = (start_x, y, start_z)
        
        # [CẢI TIẾN] Bắt đầu đường đi từ vị trí xuất phát
        all_path_coords: list[Coord] = [start_pos]
        island_placement_coords: list[Coord] = [] # [SỬA] Danh sách riêng cho tọa độ của các đảo
        
        # Vòng lặp để tạo ra các hòn đảo
        for i in range(num_islands):
            # Tọa độ góc của mỗi hòn đảo được tính toán dựa trên khoảng cách
            # [CẢI TIẾN] Nếu là đảo đầu tiên, góc chính là start_pos. Nếu không, tính toán.
            island_corner = (start_x + i * island_spacing, y, start_z) if i > 0 else start_pos
            
            # Tạo đường đi cho hòn đảo hiện tại
            island_path, _ = self._create_island_pattern(island_corner, island_pattern_name)
            island_placement_coords.extend(island_path) # Thêm vào danh sách placement
            
            # [CẢI TIẾN] Nối đường đi một cách liền mạch
            # Bỏ qua điểm đầu của island_path nếu nó đã có trong all_path_coords
            for p in island_path:
                if p not in all_path_coords:
                    all_path_coords.append(p)
            
            # Tạo "cây cầu" nối giữa các đảo (nếu không phải đảo cuối cùng)
            if i < num_islands - 1:
                # [CẢI TIẾN] Cầu nối liền mạch từ điểm cuối của đảo hiện tại
                bridge_start = all_path_coords[-1]
                bridge_end = (start_x + (i + 1) * island_spacing, y, start_z) # Điểm đầu của đảo tiếp theo
                
                # [SỬA LỖI] Logic tạo cầu mới để đảm bảo tính liên tục.
                # Bước 1: Đi từ điểm cuối của đảo (bridge_start) về lại trục Z chính (start_z).
                # Đi ngược theo trục Z.
                current_z = bridge_start[2]
                while current_z > start_z:
                    current_z -= 1
                    all_path_coords.append((bridge_start[0], y, current_z))
                
                # Bước 2: Đi từ vị trí hiện tại trên trục Z chính đến điểm bắt đầu của đảo tiếp theo.
                # Đi thẳng theo trục X.
                current_x = bridge_start[0]
                while current_x < bridge_end[0]:
                    current_x += 1
                    all_path_coords.append((current_x, y, start_z))


        target_pos = all_path_coords[-1]
        
        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=all_path_coords,
            placement_coords=all_path_coords # [FIX] Phải chứa cả đảo và cầu nối để render nền đất đầy đủ
        )

File: packages/academic-map-generator/src/generator/topologies/SymmetricalIslands.ts
/**
 * Symmetrical Islands Topology (PORTED FROM PYTHON)
 * Creates identical island structures connected by bridges.
 * Ideal for teaching about pattern recognition and function reuse.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

export class SymmetricalIslandsTopology extends BaseTopology {
  /**
   * Create an island pattern at the given corner.
   * Supports: 'u_shape', 'l_shape', 'plus_shape', 'square_shape'
   */
  private _createIslandPattern(topLeftCorner: Coord, patternName: string): { 
    path: Coord[]; 
    size: [number, number] 
  } {
    const [x, y, z] = topLeftCorner;
    let path: Coord[] = [];
    let size: [number, number] = [0, 0];

    switch (patternName) {
      case 'l_shape':
        // L-shape, 3x3
        path = [
          [x, y, z], [x, y, z + 1], [x, y, z + 2],
          [x + 1, y, z + 2], [x + 2, y, z + 2]
        ];
        size = [3, 3];
        break;
      case 'plus_shape':
        // Plus shape, 3x3
        path = [
          [x + 1, y, z],
          [x, y, z + 1], [x + 1, y, z + 1], [x + 2, y, z + 1],
          [x + 1, y, z + 2]
        ];
        size = [3, 3];
        break;
      case 'square_shape':
        // Hollow square, 3x3
        path = [
          [x, y, z], [x + 1, y, z], [x + 2, y, z],
          [x + 2, y, z + 1], [x + 2, y, z + 2],
          [x + 1, y, z + 2], [x, y, z + 2], [x, y, z + 1]
        ];
        size = [3, 3];
        break;
      case 'u_shape':
      default:
        // U-shape, 2x2
        path = [
          [x, y, z], [x + 1, y, z],
          [x + 1, y, z + 1], [x, y, z + 1]
        ];
        size = [2, 2];
        break;
    }

    return { path, size };
  }

  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;
    
    // Loop through possible island counts
    for (let numIslands = 2; numIslands <= 5; numIslands++) {
      if (count >= maxVariants) return;
      yield this.generatePathInfo(gridSize, { ...params, num_islands: numIslands });
      count++;
    }
  }

  generatePathInfo(gridSize: [number, number, number], params: Record<string, any>): IPathInfo {
    console.log("    LOG: Generating 'symmetrical_islands' topology...");

    // Parse num_islands (supports both number and range array)
    let numIslands: number;
    const numIslandsParam = params.num_islands || 2;
    if (Array.isArray(numIslandsParam) && numIslandsParam.length === 2) {
      numIslands = Math.floor(numIslandsParam[0] + Math.random() * (numIslandsParam[1] - numIslandsParam[0] + 1));
    } else {
      numIslands = numIslandsParam;
    }

    // Get island pattern
    const islandPatternName = params.island_pattern || 'u_shape';

    // Calculate island size
    const { size: islandSize } = this._createIslandPattern([0, 0, 0], islandPatternName);
    const islandWidth = islandSize[0];
    const islandDepth = islandSize[1];

    // Calculate spacing
    let islandSpacing: number;
    if ('island_gap' in params) {
      const islandGap = parseInt(params.island_gap);
      islandSpacing = islandWidth + islandGap;
    } else {
      const totalIslandsWidth = numIslands * islandWidth;
      const availableSpacing = gridSize[0] - totalIslandsWidth - 2;
      islandSpacing = numIslands > 1 
        ? islandWidth + Math.floor(availableSpacing / (numIslands - 1))
        : 0;
    }

    // Calculate safe start position
    const startX = params.start_x || 1;
    const startZ = params.start_z || Math.max(1, Math.floor(Math.random() * (gridSize[2] - islandDepth - 2)) + 1);
    const y = 0;

    const startPos: Coord = [startX, y, startZ];
    const allPathCoords: Coord[] = [startPos];
    const islandPlacementCoords: Coord[] = [];
    const islandsList: Coord[][] = [];
    const bridges: Coord[][] = [];

    // Create islands
    for (let i = 0; i < numIslands; i++) {
      // Calculate island corner position
      const islandCorner: Coord = i > 0 
        ? [startX + i * islandSpacing, y, startZ]
        : startPos;

      // Create island path
      const { path: islandPath } = this._createIslandPattern(islandCorner, islandPatternName);
      islandPlacementCoords.push(...islandPath);
      islandsList.push(islandPath);

      // Add island path to all_path_coords (avoiding duplicates)
      for (const p of islandPath) {
        if (!allPathCoords.some(c => c[0] === p[0] && c[1] === p[1] && c[2] === p[2])) {
          allPathCoords.push(p);
        }
      }

      // Create bridge to next island (if not last island)
      if (i < numIslands - 1) {
        const bridgeStart = allPathCoords[allPathCoords.length - 1];
        const bridgeEndX = startX + (i + 1) * islandSpacing;
        const bridgeCoords: Coord[] = [];

        // Step 1: Go back to base Z (start_z) along Z axis
        let currentZ = bridgeStart[2];
        while (currentZ > startZ) {
          currentZ--;
          const coord: Coord = [bridgeStart[0], y, currentZ];
          allPathCoords.push(coord);
          bridgeCoords.push(coord);
        }

        // Step 2: Go forward along X axis
        let currentX = bridgeStart[0];
        while (currentX < bridgeEndX) {
          currentX++;
          const coord: Coord = [currentX, y, startZ];
          allPathCoords.push(coord);
          bridgeCoords.push(coord);
        }

        bridges.push(bridgeCoords);
      }
    }

    const targetPos = allPathCoords[allPathCoords.length - 1];

    // Semantic positions (matching Python)
    const semantic_positions = {
      start: startPos,
      end: targetPos,
      optimal_start: 'start',
      optimal_end: 'end',
      valid_pairs: [
        {
          name: 'identical_islands_easy',
          start: 'start',
          end: 'end',
          path_type: 'island_traversal',
          strategies: ['island_replication', 'function_reuse'],
          difficulty: 'EASY',
          teaching_goal: 'Identical items per island'
        },
        {
          name: 'varied_island_sizes_medium',
          start: 'start',
          end: 'end',
          path_type: 'progressive_islands',
          strategies: ['island_replication', 'symmetric_traversal'],
          difficulty: 'MEDIUM',
          teaching_goal: 'Pattern varies by island'
        },
        {
          name: 'cross_island_pattern_hard',
          start: 'start',
          end: 'end',
          path_type: 'cross_island',
          strategies: ['island_replication', 'function_reuse'],
          difficulty: 'HARD',
          teaching_goal: 'Cross-island pattern recognition'
        }
      ]
    };

    // Segment analysis
    const allSegments = [...islandsList, ...bridges];
    const lengths = allSegments.map(s => s.length);
    const segment_analysis = {
      num_segments: allSegments.length,
      lengths,
      types: [...islandsList.map(() => 'island'), ...bridges.map(() => 'bridge')],
      min_length: lengths.length > 0 ? Math.min(...lengths) : 0,
      max_length: lengths.length > 0 ? Math.max(...lengths) : 0,
      avg_length: lengths.length > 0 ? lengths.reduce((a, b) => a + b, 0) / lengths.length : 0
    };

    // Deduplicate placement coords (all walkable tiles)
    const dedupedPlacement = this._deduplicateCoords(allPathCoords);
    
    // Compute path_coords using BFS pathfinding
    const computedPath = this.computePathCoords(startPos, targetPos, dedupedPlacement);

    return {
      start_pos: startPos,
      target_pos: targetPos,
      path_coords: computedPath,          // DYNAMIC: shortest path
      placement_coords: dedupedPlacement, // STATIC: all walkable tiles
      obstacles: [],
      metadata: {
        topology_type: 'symmetrical_islands',
        islands: islandsList,
        bridges: bridges,
        num_islands: numIslands,
        island_pattern: islandPatternName,
        segments: allSegments,
        segment_analysis,
        semantic_positions
      }
    };
  }
}
